src/widgets/lv_btnmatrix.h

```diff
@@ -39,6 +39,7 @@ enum {
     LV_BTNMATRIX_CTRL_CHECKABLE  = 0x0040, /**< The button can be toggled.*/
     LV_BTNMATRIX_CTRL_CHECKED    = 0x0080, /**< Button is currently toggled (e.g. checked).*/
     LV_BTNMATRIX_CTRL_CLICK_TRIG = 0x0100, /**< 1: Send LV_EVENT_VALUE_CHANGE on CLICK, 0: Send LV_EVENT_VALUE_CHANGE on PRESS*/
+    LV_BTNMATRIX_CTRL_NO_POPOVER = 0x0200, /**< Do not show a popover when pressing this key*/
```

Comment:
The default should be non popover, so I'd change it to `LV_BTNMATRIX_CTRL_POPOVER`.

Reply:
Hm, are you certain? Even though popovers are disabled by default, I set up the control flags on the default keymap so that popovers work out of the box on the relevant keys as soon as you do `lv_btnmatrix_set_popovers(keyboard, true)`.

Right now it looks like this:

```
static const char * const default_kb_map_lc[] = {"1#", "q", "w", "e", "r", "t", "y", "u", "i", "o", "p", LV_SYMBOL_BACKSPACE, "\n",
    ...
};

static const lv_btnmatrix_ctrl_t default_kb_ctrl_lc_map[] = {
    LV_KEYBOARD_CTRL_BTN_FLAGS | LV_BTNMATRIX_CTRL_NO_POPOVER | 5, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,  
    ...
};
```

Only the few keys that should not pop up received the `LV_BTNMATRIX_CTRL_NO_POPOVER` and everything else stays as is.

When we flip the semantic of the flag, we'd have to change the map to e.g.

```
static const lv_btnmatrix_ctrl_t default_kb_ctrl_lc_map[] = {
    LV_KEYBOARD_CTRL_BTN_FLAGS | 5, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4, LV_BTNMATRIX_CTRL_POPOVER | 4,  
    ...
};
```

That might make the default map look overly complicated.

Or did you want to totally disable popovers in the default map, meaning users should call `lv_btnmatrix_set_popovers(keyboard, true)` and also provide a custom control map with `LV_BTNMATRIX_CTRL_POPOVER` set wherever they want it?

Reply:
Hm, on the other hand, if we flip the logic to an `LV_BTNMATRIX_CTRL_POPOVER` option, that would make it easier to marry it with `LV_BTNMATRIX_CTRL_CLICK_TRIG`. I think keys that _do_ show a popover should always trigger the event on release because if the key is already sent while you look at the popover, then the popover isn't very helpful. So we could just say that setting `LV_BTNMATRIX_CTRL_POPOVER` implies `LV_BTNMATRIX_CTRL_CLICK_TRIG` being set as well.

Only issue, as said above, is that the default control map would get a bit length. :thinking:

Reply:
Ok, I updated the PR to use `LV_BTNMATRIX_CTRL_POPOVER`, made it imply `LV_BTNMATRIX_CTRL_CLICK_TRIG` and updated the default keyboard control maps.

Please let me know if this is ok and I'll update the documentation accordingly.

Reply:
> Hm, are you certain? 

Good question! :smile: It seems `LV_BTNMATRIX_CTRL_NO_POPOVER` actually means if the certain button is "popover compatible" or not. And if we don't want popovers by default we can set `lv_btnmatrix_set_popovers(btnm ,false)`.

But somehow this approach is still strange to me. Let's talk about just a plain btnmatrix. When the user sets the control map he can decide whether he wants popover on a certain button or not. So there is no need for `lv_btnmatrix_set_popovers()`. 

It's less elegant but it looks more logical to me have an `lv_keyboard_set_popover(kb, enable)` save the `enable` flag in the keyboard, and in `lv_keyboard_set_mode` add  `LV_BTNMATRIX_CTRL_POPOVER` flag to the relevant buttons if the popover is enabled. Maybe rather with `lv_btnmatrix_set_ctrl_all` to add popover to all buttons, and remove from a few. 

What do you think?

Reply:
> Let's talk about just a plain btnmatrix. When the user sets the control map he can decide whether he wants popover on a certain button or not. So there is no need for lv_btnmatrix_set_popovers().

Ah, I think I see what you're saying. The presence (or absence) of ` LV_BTNMATRIX_CTRL_POPOVER` in the control map already says that popovers are enabled (or disabled) so there is no need for a separate `lv_btnmatrix_set_popovers(btnm, true / false)` call.

Instead, we could just derive the value of `btnm->popovers` by looping over the control map and checking if at least one button has `LV_BTNMATRIX_CTRL_POPOVER` set.

> It's less elegant but it looks more logical to me have an lv_keyboard_set_popover(kb, enable) save the enable flag in the keyboard, and in lv_keyboard_set_mode add LV_BTNMATRIX_CTRL_POPOVER flag to the relevant buttons if the popover is enabled. Maybe rather with lv_btnmatrix_set_ctrl_all to add popover to all buttons, and remove from a few.

Hm, I am not sure if that would work generically. The caller could have set a custom map and control map via `lv_keyboard_set_map`. I'm actually doing that in the apps I'm working on to switch between e.g. a French and a German keyboard layout. So with a custom map, when `lv_keyboard_set_mode` is called, I believe, we wouldn't have a way to know what buttons to automatically add the popover flag to.

We could maybe tell if a custom map was set or not by checking if `kb_map[mode]` still points to the default map. Did you mean that?

Alternatively, I think what would work generically is if we inverted the logic you suggest:

1. Add `LV_BTNMATRIX_CTRL_POPOVER` in the default control maps (like currently done in the PR)
2. In `lv_keyboard_set_mode` *remove* the popover flag from all buttons if popovers were previously disabled by calling `lv_keyboard_set_popover(kb, false)`.

That would still have the disadvantage that the default control maps are bloated up a bit by the various `LV_BTNMATRIX_CTRL_POPOVER` flags but maybe that's not too bad.

Let me know what you think and thanks a ton for taking the time to talk this through. :slightly_smiling_face:

Reply:
> Alternatively, I think what would work generically is if we inverted the logic you suggest:

You are right, and it sounds good! :slightly_smiling_face: 
The bloating could be slightly avoided by adding a helper macro like 
```c
#define LV_KB_BTN(width) LV_BTNMATRIX_CTRL_POPOVER | width
```

Reply:
> However, one thing I noticed is that without a dedicated btnm->popovers setting, we now have to recompute the extended draw size any time the control map is changed and that means looping over all buttons in the top row and checking if any of them has the LV_BTNMATRIX_CTRL_POPOVER.

IMO it's acceptable to re-implement a custom `clear_all` method in the keyboard for this specific case.

Reply:
Ok, I tried that but then decided to go a slightly different road. Instead in `lv_keyboard.c` I now copy the control map, remove the `LV_BTNMATRIX_CTRL_POPOVER` on the copied map and finally hand it down to the button matrix. The benefits of this are:

1. The keyboard doesn't need to know about extended draw size recalculations or button invalidations
2. The control map is only set a single time so the extended draw size calculation also only happens once

The drawback is that we have to make a temporary copy of the control map but I think that should be cheap because it's just an integer array.

Let me know what you think. I also updated the docs so everything should be in place now.